### PR TITLE
[REF] CRM_Utils_Recent - Use hook listener to delete items

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Simplifies a bunch of delete functions by de-coupling BAOs from the Recent Items menu/sidebar.

Before
----------------------------------------
BAO functions responsible for deleting their own items from the Recent menu/sidebar.

After
----------------------------------------
Recent items delete themselves.

Technical Details
----------------------------------------
This is toward [dev/core#2757](https://lab.civicrm.org/dev/core/-/issues/2757)

Comments
----------------------------------------
All that's missing is `Note` which has a weird `del()` function. That's addressed in #21208

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@colemanw I just skimmed through this & obviously my take was that it really should have a test - but I know you've done a lot of PRs in this space just recently so wondering if you feel that you have otherwise covered it (or perhaps there already is a test)

**Comment by colemanw** [comment]
I don't think there are any tests at all for `CRM_Utils_Recent`!

**Comment by colemanw** [comment]
@eileenmcnaughton and now there is one :)

**Comment by eileenmcnaughton** [comment]
PHP Fatal error:  Cannot declare class api\v4\Action\ContactIsDeletedTest, because the name is already in use in /home/jenkins/bknix-dfl/build/core-21204-7ayeg/web/sites/all/modules/civicrm/tests/phpunit/api/v4/Action/ContactIsDeletedTest.php on line 81
PHP Stack trace:

File: CRM/Activity/BAO/Activity.php [hunk 1]
```diff
@@ -217,15 +217,6 @@ public static function deleteActivity(&$params, $moveToTrash = FALSE) {
       self::logActivityAction($activity, $logMsg);
     }
 
-    // delete the recently created Activity
```
Comment (eileenmcnaughton) [inline CRM/Activity/BAO/Activity.php:220]: I just r-run tested this one & it worked (although the recent items block does require a screen refresh)

File: CRM/Case/BAO/Case.php [hunk 2]
```diff
@@ -217,12 +217,6 @@ public static function deleteCase($caseId, $moveToTrash = FALSE) {
 
       CRM_Utils_Hook::post('delete', 'Case', $caseId, $case);
 
-      // remove case from recent items.
```
Comment (eileenmcnaughton) [inline CRM/Case/BAO/Case.php:220]: r-run ++

File: CRM/Contact/BAO/Contact.php [hunk 3]
```diff
@@ -1063,7 +1063,7 @@ public static function deleteContact($id, $restore = FALSE, $skipUndelete = FALS
     }
 
     //delete the contact id from recently view
-    CRM_Utils_Recent::delContact($id);
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Contact.php:1066]: Yep - tested all good - note that when soft deleting it gets a strike through & is removed on hard-delete

File: CRM/Contact/BAO/Group.php [hunk 4]
```diff
@@ -116,13 +116,6 @@ public static function discard($id) {
     $transaction->commit();
 
     CRM_Utils_Hook::post('delete', 'Group', $id, $group);
-
-    // delete the recently created Group
-    $groupRecent = [
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Group.php:121]: r-run = OK

File: CRM/Grant/BAO/Grant.php [hunk 5]
```diff
@@ -259,13 +259,6 @@ public static function del($id) {
 
     $grant->find();
 
-    // delete the recently created Grant
-    $grantRecent = [
```
Comment (eileenmcnaughton) [inline CRM/Grant/BAO/Grant.php:263]: r-run = ok

**Comment by eileenmcnaughton** [comment]
I did r-run on a handful of entities here and also stepped through the code. I thought it was great that you covered permissions in the test because I didn't think of that when doing r-run.

It would be great to also CREATE the r-run entities in the post function but obviously that is scope creep!

**Pull request merged**